[SPARK-5801] [core] Avoid creating nested directories.

vanzin opened the pull request:
Cache the value of the local root dirs to use for storing local data,
so that the same directories are reused.

Also, to avoid an extra level of nesting, use a different env variable
to propagate the local dirs from the Worker to the executors. And make
the executor directory use a different name.

Commit:
[SPARK-5801] [core] Avoid creating nested directories.

Cache the value of the local root dirs to use for storing local data,
so that the same directory is reused.

Also, to avoid an extra level of nesting, use a different env variable
to propagate the local dirs from the Worker to the executors. And make
the executor directory use a different name.

SparkQA commented:
[Test build #27903 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27903/consoleFull) for   PR 4747 at commit [`18ee0a7`](https://github.com/apache/spark/commit/18ee0a7435f9048dde1d79373891703a1187e584).
 * This patch merges cleanly.

srowen commented on core/src/main/scala/org/apache/spark/deploy/worker/ExecutorRunner.scala:
```diff
@@ -135,7 +135,7 @@ private[spark] class ExecutorRunner(
 
       builder.directory(executorDir)
-      builder.environment.put("SPARK_LOCAL_DIRS", appLocalDirs.mkString(","))
+      builder.environment.put("SPARK_EXECUTOR_DIRS", appLocalDirs.mkString(File.pathSeparator))
```
I think these are dumb questions, but why does the fix entail setting a different env variable? and should dirs be joined with a path separator?

vanzin replied:
Without this, `Utils.scala` would always create a subdirectory for the executor under the directory already created by the Worker. So you'd always have "spark-xxxxx/spark-yyyyy" for every executor. It would be always like that (no extra nesting), but, since I'm changing this, sounded like a good enhancement.

And dirs should be joined with a path separator because that's the right way to do it.

srowen replied:
Oh right, this is `path.separator` not `file.separator`, `:` not `/`. 

Yes that's exactly the fix that needs to be made.

`SPARK_EXECUTOR_DIRS` is a new system property right? where is it read? I am missing something.

srowen commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -679,12 +680,25 @@ private[spark] object Utils extends Logging {
    * If no directories could be created, this will return an empty list.
    */
   private[spark] def getOrCreateLocalRootDirs(conf: SparkConf): Array[String] = {
+    if (localRootDirs == null) {
```
Can this whole mechanism be replaced with a `lazy val` for `localRootDirs`?

vanzin replied:
The problem with the lazy val is that you don't get access to the `SparkConf` argument.

vanzin replied:
(On a side note, this patch might do some weird things if the method is called with different confs.)

vanzin commented:
Hold off a little bit on this; let me see how to handle the SparkConf issue (I think Josh added that for the unit tests).

SparkQA commented:
[Test build #27903 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27903/consoleFull) for   PR 4747 at commit [`18ee0a7`](https://github.com/apache/spark/commit/18ee0a7435f9048dde1d79373891703a1187e584).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27903/
Test FAILed.

andrewor14 commented:
retest this please. We really need to fix the `JavaAPISuite` at some point.

SparkQA commented:
[Test build #27908 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27908/consoleFull) for   PR 4747 at commit [`18ee0a7`](https://github.com/apache/spark/commit/18ee0a7435f9048dde1d79373891703a1187e584).
 * This patch merges cleanly.

Commit:
Update unit test.

Add a way for the test to work around the cache (even though the test
was passing without this). Also add a comment that explains the new
behavior of the `getOrCreateLocalRootDirs()`.

SparkQA commented:
[Test build #27909 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27909/consoleFull) for   PR 4747 at commit [`e0114e1`](https://github.com/apache/spark/commit/e0114e19a8d3ae5cb8aa4d253f263f1407fdfb21).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27908 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27908/consoleFull) for   PR 4747 at commit [`18ee0a7`](https://github.com/apache/spark/commit/18ee0a7435f9048dde1d79373891703a1187e584).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27908/
Test PASSed.

SparkQA commented:
[Test build #27909 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27909/consoleFull) for   PR 4747 at commit [`e0114e1`](https://github.com/apache/spark/commit/e0114e19a8d3ae5cb8aa4d253f263f1407fdfb21).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27909/
Test PASSed.

Sephiroth-Lin commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -728,6 +746,11 @@ private[spark] object Utils extends Logging {
   }
 
+  /** Used by unit tests. Do not call from other places. */
+  private[spark] def clearLocalRootDirs(): Unit = {
```
May be we can call this function to delete the local root directory in non-yarn mode when application is exited or SparkContext is stoped.

vanzin replied:
Note this method doesn't delete anything. There's a separate PR to clean up the root directories (#4759).

srowen commented on core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:
```diff
@@ -345,11 +345,11 @@ private[spark] class Worker(
           val appLocalDirs = appDirectories.get(appId).getOrElse {
             Utils.getOrCreateLocalRootDirs(conf).map { dir =>
-              Utils.createDirectory(dir).getAbsolutePath()
+              Utils.createDirectory(dir, namePrefix = "executor").getAbsolutePath()
```
Do we need to register these for cleanup at shutdown? cf. https://github.com/apache/spark/pull/4759#issuecomment-75937545

vanzin replied:
Not needed (see SPARK-4834). That's done by the `maybeCleanupApplication` in this file when the application finishes.

srowen commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -63,6 +63,7 @@ private[spark] object Utils extends Logging {
   val random = new Random()
 
   private val MAX_DIR_CREATION_ATTEMPTS: Int = 10
+  @volatile private var localRootDirs: Array[String] = null
```
OK at first I was uneasy about adding more global state to a common `Utils` class, but this reflects configuration that's already global to the whole JVM, coming by sys properties like `SPARK_LOCAL_DIRS`.  Maybe this could be declared nearer its usage but that alone isn't worth changing this. I also don't see a way to avoid the double-checked locking; I couldn't get something clean with `AtomicReference` to work. I think the extra `SPARK_EXECUTOR_DIRS` is worth it; it's not exposed otherwise.

In other words, LGTM. Any other reviews? I'll merge after waiting a short while for comments.
